### PR TITLE
WIP: fix bound evaluation in dist_math

### LINKED ISSUE #1579: Error when using DiscreteUniform
When reproducing http://isaacslavitt.com/2015/12/19/german-tank-problem-with-pymc-and-pystan/

With the newest rc4 - I run into the following ridiculous results when running `pm.summary(trace)`

```

N:

  Mean             SD               MC Error         95% HPD interval
  -------------------------------------------------------------------
  
  2687.276         2570.593         25.819           [256.000, 8376.000]

  Posterior quantiles:
  2.5            25             50             75             97.5
  |--------------|==============|==============|--------------|
  
  280.000        642.000        1643.000       4074.000       9145.000
```

These are off by a factor of 10. 

Anyone any ideas? I used the same code.

### PR DESCRIPTION
**Pull request by fonnesbeck**
This is a first pass at fixing the bug in `bound` that manifests itself in #1579. I've used stack to get around the heterogeneous arguments, but this results in dimension-matching problems for some models that I don't yet understand.

Also added the tanks example from #1579 to `test_examples`.

Closes #1579

### CODE REVIEW

**Comment by twiecki** [comment]
does the original code not work?

**Comment by fonnesbeck** [comment]
Original code?

**Comment by twiecki** [comment]
```python
     ret = 1
     for c in vals:
         ret = ret * (1 * c)
     return ret
```

**Comment by twiecki** [comment]
https://github.com/pymc-devs/pymc3/commit/1b7b4bb2c5b49ea92794d86285b2c5086af7ed99

**Comment by fonnesbeck** [comment]
I assumed it would reintroduce #1449

**Comment by twiecki** [comment]
Oh that was the original issue.

I think that bears taking another look. Either we have broadcasting in `bound` which can lead to sometimes odd behavior as in #1449 or we don't, but I don't think there's a middle ground.

**Comment by twiecki** [comment]
Perhaps we should have a kwarg `broadcast=False` that can turn broadcasting on or off. Both use-cases (DiscreteUniform and Multinomial) are valid I think.

**Comment by twiecki** [comment]
How does this handle the #1449 case?

**Comment by fonnesbeck** [comment]
It does fix it, at least based on the example @AustinRochford used in the issue. Perhaps I will include that as a test, incase of regression.

**Comment by fonnesbeck** [comment]
Please have a peek at this @ferrine and/or @AustinRochford

**Comment by ferrine** [comment]
I'm a bit sick and have a deadline for tomorrow:( I'll review the PR in some days

**Comment by fonnesbeck** [comment]
@ferrine no big deal, thanks. Get well.

**Comment by AustinRochford** [comment]
This looks like a reasonable compromise to me. 👍

File: pymc3/distributions/dist_math.py [hunk 1]
```diff
@@ -29,8 +29,16 @@ def bound(logp, *conditions):
...
 def alltrue(vals):
-    return tt.all([tt.all(1 * val) for val in vals])
+    """
+    Asserts truth of all elements in vals, across the lowest axis. This maintains
+    element-wise evaluations for multivariate inputs.
+    """
+    try:
+        return tt.all(tt.stack([1*val for val in vals]), axis=0)
+    except (TypeError, IndexError):
+        return tt.all([tt.all(1 * val) for val in vals])
```
Comment (jsalvatier) [inline pymc3/distributions/dist_math.py:39]: When is it right to do this second version?

Comment (fonnesbeck) [reply pymc3/distributions/dist_math.py:39]: Sometimes `tt.stack` will fail due to different lengths of the elements of `vals` (the TypeError; happens with the Poisson mixture) and when there is no axis to iterate over (resulting in the IndexError).

Comment (ferrine) [reply pymc3/distributions/dist_math.py:39]: You can try to ravel all variables after safe converting them in theano, then concatenate and finally use tt.all

Comment (fonnesbeck) [reply pymc3/distributions/dist_math.py:39]: I thought stack did most of that (including the conversion) for me automatically?

Comment (fonnesbeck) [reply pymc3/distributions/dist_math.py:39]: You run into situations where `vals` is not only of different types, but also of different lengths, so we need to deal with those potential combinations. `tt.stack` doesn't like heterogeneous dimensions. Haven't seen a case that tricks both yet, however.

Comment (ferrine) [reply pymc3/distributions/dist_math.py:39]: Different dims can be excluded by ravel

**Comment by jsalvatier** [comment]
hmm, they should always be broadcastable together though (since this is
elementwise). It seems to me that it should work like this.

Actually, I think I'm not clear why we're not sticking with my original
version. What exactly is the extra case that we want to handle?

This version will sometimes collapse down the conditions into a single
value, which is not what we want for elementwise. We do want that (or are
ok with it) for non-elementwise distributions, but we should just have a
different function for that. Or make a more explicit switch of some kind.
Right now, you might accidentally get the collapsing behavior, and that
will be confusing.

Stacking doesn't handle arrays that can be broadcasted to together but
aren't the same shape (some of the inputs to conditions may not be the full
shape), which we need for this.


On Mon, Dec 12, 2016 at 12:57 PM Chris Fonnesbeck <notifications@github.com>
wrote:

*@fonnesbeck* commented on this pull request.
------------------------------

In pymc3/distributions/dist_math.py
<https://github.com/pymc-devs/pymc3/pull/1591>:

> @@ -29,8 +29,16 @@ def bound(logp, *conditions):


 def alltrue(vals):
-    return tt.all([tt.all(1 * val) for val in vals])
+    """
+    Asserts truth of all elements in vals, across the lowest axis.
This maintains
+    element-wise evaluations for multivariate inputs.
+    """
+    try:
+        return tt.all(tt.stack([1*val for val in vals]), axis=0)
+    except (TypeError, IndexError):
+        return tt.all([tt.all(1 * val) for val in vals])

You run into situations where vals is not only of different types, but also
of different lengths, so we need to deal with those potential combinations.
tt.stack doesn't like heterogeneous dimensions. Haven't seen a case that
tricks both yet, however.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/pymc-devs/pymc3/pull/1591>, or mute the thread

**Comment by fonnesbeck** [comment]
The original problem is #1449, which had some unexpected broadcasting behavior.

**Comment by jsalvatier** [comment]
Right, okay. Then suggest we make `bound_elemwise` and `bound` and `alltrue_elemwise` and `alltrue` or do something similar.

**Comment by fonnesbeck** [comment]
Perhaps its just simpler (or at least clearer, if not simpler) to call  `all` explicitly in each distribution as needed, and pass that to `bound`. e.g. `tt.all(value >= 0)`, `tt.all(0 <= p)`, etc. That way, each call to the `logp` would always return a 1D array of booleans (one for each condition).

**Comment by jsalvatier** [comment]
For multivariate distributions? That would be okay. Though I think we should still rename `bound` to `elemwise_bound` or similar. Just to be less confusing. 

We should keep elemwise bound to be completely elemwise so that the logps are elemwise (and individual elements can be nans). Its useful for debugging.

**Comment by fonnesbeck** [comment]
No, I mean for all the distributions (multivariates already do this). Unless I am missing something, if any condition for any element of a call to `logp` fails, the `logp` should return `-inf`. You can still debug inside of the `logp` if you need to. 

So, if you have a vector-valued Gamma distribution, for example, you would return `-inf` on the call to `logp` on any violation of the bound by any element, be it a bad parameter element or a bad value element.

**Comment by jsalvatier** [comment]
Gotcha, I think that seems bad to me. Elementwise distributions should be as elementwise as possible. This change would make them partially a single distribution. Its more conceptually coherent.

What do you mean you can still debug inside the logp? 

The problems I'm thinking of are things like: if you pass bad initial values to an elemwise dist, its better if you can look at the logp to tell which ones are bad. Or if find_MAP messes up. Or if you write a bad sampler or distribution. 

You can also imagine custom samplers using the -inf information to adjust the scale for particular elements. 

I also don't really see the benefit. We can just have two functions with clearer names.

**Comment by fonnesbeck** [comment]
I think I've lost the plot on this one. I will close this and let someone else take a shot at it. I'm sure the answer is easy and I'm just missing it.

**Pull request closed without merging**